### PR TITLE
REF: implement Dtype.index_class

### PR DESCRIPTION
**Pull request by jbrockmendel**
Allows EAs to specify what Index subclass they want to use.  Avoids special-casing pandas-internal EAs.

Small perf improvement

```
In [2]: dti = pd.date_range("2016-01-01", periods=3, tz="UTC")

In [3]: cat = dti.astype("category")

In [4]: %timeit pd.Index(dti)
6.39 µs ± 80 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each) # <- main
5.69 µs ± 368 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)  # <- PR

In [5]: %timeit pd.Index(cat)
6.37 µs ± 112 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)  # <- main
5.8 µs ± 135 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)  # <- PR

```

### CODE REVIEW

File: pandas/core/dtypes/dtypes.py [hunk 1]
```diff
@@ -671,6 +675,12 @@ def _get_common_dtype(self, dtypes: list[DtypeObj]) -> DtypeObj | None:
 
         return find_common_type(non_cat_dtypes)
 
+    @cache_readonly
```
Comment (mroeschke) [inline pandas/core/dtypes/dtypes.py:678]: Could you make these properties like the base class?

Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:678]: updated to make both cache_readonly; I think the perf improvement in the OP depends on it

Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:678]: Updated + greenish

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**